File: 2025-03-March-GFG-POTD/March 01 - Decode the string/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string decodeString(string s) {
        stack<int> countStack;
        stack<string> stringStack;
        string currentString = "";
        int num = 0;
        
        for (char ch : s) {
            if (isdigit(ch)) {
                num = num * 10 + (ch - '0'); // Handle multi-digit numbers
            } else if (ch == '[') {
                countStack.push(num);
                stringStack.push(currentString);
                currentString = "";
                num = 0;
            } else if (ch == ']') {
                int repeatCount = countStack.top(); countStack.pop();
                string decodedString = stringStack.top(); stringStack.pop();
                while (repeatCount-- > 0) {
                    decodedString += currentString;
                }
                currentString = decodedString;
            } else {
                currentString += ch;
            }
        }
        
        return currentString;
    }
};

// Driver Code
int main() {
    int t;
    cin >> t;
    while (t--) {
        string s;
        cin >> s;
        Solution obj;
        cout << obj.decodeString(s) << endl;
    }
    return 0;
}

File: 2025-03-March-GFG-POTD/March 01 - Decode the string/Solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        count_stack = []
        string_stack = []
        current_string = ""
        num = 0
        
        for ch in s:
            if ch.isdigit():
                num = num * 10 + int(ch)  # Handle multi-digit numbers
            elif ch == "[":
                count_stack.append(num)
                string_stack.append(current_string)
                current_string = ""
                num = 0
            elif ch == "]":
                repeat_count = count_stack.pop()
                decoded_string = string_stack.pop()
                current_string = decoded_string + current_string * repeat_count
            else:
                current_string += ch
        
        return current_string

# Driver Code
if __name__ == "__main__":
    t = int(input().strip())
    for _ in range(t):
        s = input().strip()
        obj = Solution()
        print(obj.decodeString(s))
